### ISSUE TITLE
Json Type Provider from Json-schema 

### ISSUE THREAD

**Issue by jruizaranguren** [fsharp/FSharp.Data#677]
Json-Schema could provide us with a principled and standardized way to understand json data.
Although it has not got much traction yet is a powerfull and complete standard with [a couple of decent implementations.](http://json-schema.org/implementations.html)

It migth simplify type inference and reduce ambiguity. It also allows for shareable lightweight data specifications.

A brief brainstorming of json-schema keywords that could be used to leverage F# characteristics:
- "additionalProperties": differenciate record type vs classes.
- "oneOf": specify discriminated unions (also with "enum")
- "required": for "option"
- "maxItems/minItems": array cardinality (validation)
- "uniqueItems": set behavior in arrays.
- "multipleOf,maximum,minimum": for selection of numeric types with desirable precision.
- "maxLength,minLength,pattern": for specific string validations.

We would also have a standard way to:
- nest and reference schemas through URIs.
- richer semantics for validation of data
- types could be named (even a namespace convention could be generated from schemas).
- room for creating more performant parsers dinamically
- specify APIs, as in [heroku](https://blog.heroku.com/archives/2014/1/8/json_schema_for_heroku_platform_api) and [google discovery service](https://developers.google.com/discovery/v1/getting_started)

I'm willing to contribute to this approach, I have worked with json-schema a couple of months.

**Comment by blumu** [comment]
Jet.com has released sources for a related project: https://github.com/jet/JsonSchemaProvider

**Comment by jruizaranguren** [comment]
@blumu, it depends on Json.Net and it just conforms to JSON-Schema v3 (deprecated). Newtonsoft released a [JsonSchema library](http://www.newtonsoft.com/jsonschema), but it has license restrictions.

**Comment by tony-johnson-i2o** [comment]
Was work ever started on this? Are you still looking for someone to pick it up?

**Comment by realvictorprm** [comment]
I wonder too

**Comment by realvictorprm** [comment]
@tpetricek can we get this going again? Is it technical seen still relevant?

**Comment by tpetricek** [comment]
@realvictorprm I do not know how widely is JSON Schema adopted, but aside from that, this still sounds interesting and relevant, so if anyone wants to have a go, I'm all for it :)

**Comment by realvictorprm** [comment]
Good, need this for glTF which specification is in schema json. Can you point me a bit?

**Comment by realvictorprm** [comment]
@tpetricek I have some questions which need to be answered so I can start developing a new up to date json schema type provider:
1. Do we use an external library for parsing the schemas or should I write a complete new one? 
1.1 If I should write a new one, on which existing code should I write the parser. Do we have helper functions or restrictions on how to write the parser? (I would go with parsing the schema as regular json file and afterwards visit each field to validate and create a correct schema type)
2. Must it be an erased type provider then?
3. Is it fine to go with the latest json schema version?
4. Anything else you might want to tell me? (e.g common mistakes in contributing to this repository)
5. Should I amend my commits everytime I make small changes or is it fine to make commits for reverting and re-factoring an algorithm?

**Comment by realvictorprm** [comment]
@tpetricek feel free to assign this to me.

Tracking PR is #1069

**Comment by DoubleCouponDay** [comment]
I'm confused as to the original posters intent for a pull request. Do you wish to allow json schemas to become a JsonProvider type for multiple json files, which implement the same schema? Currently I don't think this is possible since JsonProvider still includes json-schema specific keywords, such as `Properties`.

**Comment by blumu** [comment]
@doublecouponday that's right. I meant a type provider generating types based off a JSON schema file. The suggested workaround is not meeting the requirement, in particular a single sample JSON cannot always represent all the possible JSON files accepted by a given schema.

**Comment by olstakh** [comment]
This would be so great to have! I have a list of json files that all follow the same [schema](https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#). I would be great to load JsonProvider with a schema and then parse individual files! Without it, i have to have a provider for each separate file, which is not scalable

**Comment by kurt-mueller-osumc** [comment]
> This would be a nice feature to add - and it is actually not that hard to implement!
> 
> The key thing is to parse the JSON schema and, based on the schema, generate value of the `InferedType` discriminated union ([you can see It here](https://github.com/fsharp/FSharp.Data/blob/master/src/CommonRuntime/StructuralTypes.fs#L58)), which is our internal representation of a type. Once we have that, the rest of the type provider can generate a nice .NET type from the internal representation (including things like options).
> 
> Aside from that, you'd have to add static parameter (say `Schema`) to the [list of parameters](https://github.com/fsharp/FSharp.Data/blob/master/src/Json/JsonProvider.fs#L70), and change [the code that reads the sample](https://github.com/fsharp/FSharp.Data/blob/master/src/Json/JsonProvider.fs#L47) to instead look at the schema (and produce `InferedType`)

Say I wanted to give I wanted to pick this up and start a my own attempt a pull request for this feature. Any advice from @tpetricek or @dsyme as to where to start? I was going to look at how the XML Type Provider uses XSDs to generate types.

**Comment by kurt-mueller-osumc** [comment]
> Jet.com has released sources for a related project: https://github.com/jet/JsonSchemaProvider

Does anybody have a copy of this repo? It no longer exists.

**Comment by xperiandri** [comment]
There is a newer one https://github.com/florenzen/JsonSchemaProvider
It works but it is not production-ready.
The author agrees to reuse that code to FSharp.Data.
I tried improving it and migrating it from NJsonSchema to [JsonSchema.Net](https://www.nuget.org/packages/JsonSchema.Net/) [here](https://www.nuget.org/packages/JsonSchema.Net/)
